libbpf-tools/trace_helpers.c

```diff
@@ -159,6 +166,134 @@ const struct ksym *ksyms__get_symbol(const struct ksyms *ksyms,
 	return NULL;
 }
 
+struct parts {
+	struct part *parts;
+	int parts_sz;
+	int parts_cap;
+	char *strs;
+	int strs_sz;
+	int strs_cap;
+};
+
+static int parts__add_part(struct parts *parts, const char *name,
+			   unsigned int dev)
+{
+	size_t new_cap, name_len = strlen(name) + 1;
+	struct part *part;
+	void *tmp;
+
+	if (parts->strs_sz + name_len > parts->strs_cap) {
+		new_cap = parts->strs_cap * 4 / 3;
+		if (new_cap < parts->strs_sz + name_len)
+			new_cap = parts->strs_sz + name_len;
+		if (new_cap < 64)
+			new_cap = 64;
+		tmp = realloc(parts->strs, new_cap);
+		if (!tmp)
+			return -1;
+		parts->strs = tmp;
+		parts->strs_cap = new_cap;
+	}
+	if (parts->parts_sz + 1 > parts->parts_cap) {
+		new_cap = parts->parts_cap * 4 / 3;
+		if (new_cap < 64)
+			new_cap = 64;
+		tmp = realloc(parts->parts, sizeof(*parts->parts) * new_cap);
+		if (!tmp)
+			return -1;
+		parts->parts = tmp;
+		parts->parts_cap = new_cap;
+	}
+
+	part = &parts->parts[parts->parts_sz];
+	/* while constructing, re-use pointer as just a plain offset */
+	part->name = (void *)(unsigned long)parts->strs_sz;
+	part->dev = dev;
+
+	memcpy(parts->strs + parts->strs_sz, name, name_len);
+	parts->strs_sz += name_len;
+	parts->parts_sz++;
+
+	return 0;
+}
+
+struct parts *parts__load(void)
+{
+	char part_name[DISK_NAME_LEN];
+	unsigned int devmaj, devmin;
+	unsigned long long nop;
+	struct parts *parts;
+	char buf[64];
+	FILE *f;
+	int i;
+
+	f = fopen("/proc/partitions", "r");
+	if (!f)
+		return NULL;
+
+	parts = calloc(1, sizeof(*parts));
+	if (!parts)
+		goto err_out;
+
+	while (fgets(buf, sizeof(buf), f) != NULL) {
+		/* skip heading */
+		if (buf[0] != ' ' || buf[0] == '\n')
+			continue;
+		if (sscanf(buf, "%u %u %llu %s", &devmaj, &devmin, &nop,
+				part_name) != 4)
+			goto err_out;
+		if (parts__add_part(parts, part_name, MKDEV(devmaj, devmin)))
+			goto err_out;
+	}
+
+	/* now when strings are finalized, adjust pointers properly */
+	for (i = 0; i < parts->parts_sz; i++)
+		parts->parts[i].name += (unsigned long)parts->strs;
+
+	fclose(f);
+	return parts;
+
+err_out:
+	parts__free(parts);
+	fclose(f);
+	return NULL;
+}
+
+void parts__free(struct parts *parts)
+{
+	if (!parts)
+		return;
+
+	free(parts->parts);
+	free(parts->strs);
+	free(parts);
+}
+
+const struct part *parts__map_dev(const struct parts *parts, unsigned int dev)
```

Comment:
map_dev is not coherent with get_name below. How about `partitions__get_by_dev` and `partitions__get_by_name`?

Reply:
Agree, yours are much better, thanks!